cudamapper/src/index_gpu.cuh

```diff
@@ -71,72 +73,102 @@ public:
     /// \param index_host_copy is a copy of index for a set of reads which has been previously computed and stored on the host.
     /// \param cuda_stream CUDA stream on which the work is to be done. Device arrays are also associated with this stream and will not be freed at least until all work issued on this stream before calling their destructor is done
     IndexGPU(DefaultDeviceAllocator allocator,
-             const IndexHostCopy& index_host_copy,
+             const IndexHostCopy* const index_host_copy,
              const cudaStream_t cuda_stream = 0);
 
     /// \brief returns an array of representations of sketch elements
+    /// \throw IndexNotReadyException if called before wait_to_be_ready()
     /// \return an array of representations of sketch elements
     const device_buffer<representation_t>& representations() const override;
 
     /// \brief returns an array of reads ids for sketch elements
+    /// \throw IndexNotReadyException if called before wait_to_be_ready()
     /// \return an array of reads ids for sketch elements
     const device_buffer<read_id_t>& read_ids() const override;
 
     /// \brief returns an array of starting positions of sketch elements in their reads
+    /// \throw IndexNotReadyException if called before wait_to_be_ready()
     /// \return an array of starting positions of sketch elements in their reads
     const device_buffer<position_in_read_t>& positions_in_reads() const override;
 
     /// \brief returns an array of directions in which sketch elements were read
+    /// \throw IndexNotReadyException if called before wait_to_be_ready()
     /// \return an array of directions in which sketch elements were read
     const device_buffer<typename SketchElementImpl::DirectionOfRepresentation>& directions_of_reads() const override;
 
     /// \brief returns an array where each representation is recorder only once, sorted by representation
+    /// \throw IndexNotReadyException if called before wait_to_be_ready()
     /// \return an array where each representation is recorder only once, sorted by representation
     const device_buffer<representation_t>& unique_representations() const override;
 
     /// \brief returns first occurrence of corresponding representation from unique_representations() in data arrays, plus one more element with the total number of sketch elements
+    /// \throw IndexNotReadyException if called before wait_to_be_ready()
     /// \return first occurrence of corresponding representation from unique_representations() in data arrays, plus one more element with the total number of sketch elements
     const device_buffer<std::uint32_t>& first_occurrence_of_representations() const override;
 
     /// \brief returns number of reads in input data
+    /// \throw IndexNotReadyException if called before wait_to_be_ready()
     /// \return number of reads in input data
     read_id_t number_of_reads() const override;
```

Comment:
I think we can guarantee for an index class that `number_of_reads()`, `smallest_read_id()`, `largest_read_id()` and potentially `number_of_basepairs_in_longest_read()` are callable without calling `wait_to_be_ready()`, or would that limit us in future developments in any way?
If so, let's remove the exceptions from those functions.

Reply:
I agree that some of these values can be know before device work is done because they are completely determined on host.
I would however argue that it is better to make the user call `wait_to_be_ready()` before any access for consistency as otherwise the user would have to think every time which would probably lead to more errors. It's a similar argument to the one we had for making calls to `wait_to_be_ready()` mandatory after generating constructor

Reply:
True. I am mostly concerned about paying unnecessary exception handling overhead in these trivial functions. We don't need to document that these are callable without calling `wait_to_be_ready()`. Would that be acceptable?

Reply:
I agree, but I think that overall exception handling cost is negligible compared to the amount of time spent in device code. Not documenting would be an option, but having two functions next to each other that do a similar thing where one specifies that `wait_to_be_ready()` has to be called and the other does not say anything would implicitly document it or cause confusion
As you wish, I'd be fine with both removing it and leaving it as it is

Reply:
Fair point. Let's leave it as it is.